src/bucket/BucketOutputIterator.cpp

```diff
@@ -129,14 +132,33 @@ BucketOutputIterator::getBucket(BucketManager& bucketManager,
         releaseAssert(mObjectsPut == 0);
         releaseAssert(mBytesPut == 0);
         CLOG_DEBUG(Bucket, "Deleting empty bucket file {}", mFilename);
-        std::remove(mFilename.c_str());
+        std::filesystem::remove(mFilename);
         if (mergeKey)
         {
             bucketManager.noteEmptyMergeOutput(*mergeKey);
         }
         return std::make_shared<Bucket>();
     }
-    return bucketManager.adoptFileAsBucket(mFilename, mHasher.finish(),
-                                           mObjectsPut, mBytesPut, mergeKey);
+
+    auto hash = mHasher.finish();
+    std::unique_ptr<BucketIndex const> index{};
+
+    // If this bucket needs to be indexed and is not already indexed
+    if (shouldSynchronouslyIndex)
+    {
+        // either it's a new bucket or we just reconstructed a bucket
+        // we already have, in any case ensure we have an index
+        if (auto b = bucketManager.getBucketIfExists(hash);
+            !b || !b->isIndexed())
+        {
+            std::atomic_bool exit{false};
+            index = BucketIndex::createIndex(bucketManager.getConfig(),
```

Comment:
I've changed the early exit interface to stop indexing early if the `BucketManager` is shutting down instead of relying on the passed in `atomic_bool`, so this should be resolved.

Reply:
thanks!